Feat/add custom domain

arkhoss opened the pull request:
## Description
This MR does the following:

- Support for custom domain names per stage
- Support for custom domain name mapping
- Updated `README.md` to generate documentation with `terraform-docs`
- Added `versions.tf` to add providers and requirements

## Motivation and Context
This is a common feature that should be supported, in this case it is using API Gateway v2 resources to be able to support the mapping, domain resources are the same for v1 and v2.

## Breaking Changes
No breaking changes, the creation of the domain resources and mapping can be enable disable without impact other resources in the module.

## How Has This Been Tested?
- Provisioned a real API with this changes without issues.

adamwshero commented on main.tf:
```diff
@@ -181,3 +181,33 @@ resource "aws_wafv2_web_acl_association" "this" {
+    certificate_arn                        = var.domain_certificate_arn
+    ownership_verification_certificate_arn = var.domain_ownership_verification_certificate_arn
+    endpoint_type                          = "REGIONAL"
+    security_policy                        = "TLS_1_2"
```
Do we want make variables for these or just tell users they can override the values?

```
endpoint_type  = "REGIONAL"
security_policy = "TLS_1_2"
```

adamwshero replied:
I see. The difference here is that v2 does not support `EDGE` but v1 does. We should use the v1 resource so people can make a choice.

```
[types](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_domain_name#types) - (Required) List of endpoint types. This resource currently only supports managing a single value. Valid values: EDGE or REGIONAL. If unspecified, defaults to EDGE. Must be declared as REGIONAL in non-Commercial partitions. Refer to the [documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/create-regional-api.html) for more information on the difference between edge-optimized and regional APIs.
```

arkhoss replied:
moved to use v1 resources. Thanks!

adamwshero reviewed (changes requested):
See question on domain name resource...

adamwshero commented on main.tf:
```diff
@@ -181,3 +181,33 @@ resource "aws_wafv2_web_acl_association" "this" {
+  api_id      = aws_api_gateway_rest_api.this.id
+  domain_name = aws_apigatewayv2_domain_name.this[count.index].id
+  stage       = aws_api_gateway_stage.this[count.index].stage_name
+}
```
I'd like to know more about this approach using the api_gateway_domain_name & mapping (v2) to do this instead of the v1 resources available for this.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_domain_name

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_base_path_mapping

arkhoss replied:
both works regardless the version and type of the API, imho v2 gets a bit better support for domain names than v1, but I can change it to use v1 only. Is strict requirement?

adamwshero replied:
It seems v2 doesn't support both EDGE and REGIONAL which is used in v1...

arkhoss replied:
moved to use v1 resources, thanks!

arkhoss commented:
> See question on domain name resource & domain name mapping

I have fixed endpoint type issue and moved to v1 resources. Please review and let me know if there is anything else that can be improved!
